Deployment logging slows the build tremendously

My project contains thousands of files to output (guess which project it is). The task output all these files in the console, and Kudu creates a new entry tag in its deployment log file for every new line.

```xml
        <entry time="2015-07-17T20:24:11.304537Z" id="1a9c5937-9e92-4135-a3bd-94dcaf2d959d" type="0">
            <message>  Copying file from "D:\home\site\repository\build\Compile\System.Web.WebPages.Razor.dll" to "D:\home\site\repository\build\Compile\_PublishedWebsites\Lucene\bin\System.Web.WebPages.Razor.dll".</message>
        </entry>
        <entry time="2015-07-17T20:24:11.5470539Z" id="7aabecc1-5270-4945-be6c-459f6bb85107" type="0">
            <message>  Copying file from "D:\home\site\repository\build\Compile\System.Web.Razor.dll" to "D:\home\site\repository\build\Compile\_PublishedWebsites\Lucene\bin\System.Web.Razor.dll".</message>
        </entry>
        <entry time="2015-07-17T20:24:11.8595585Z" id="9f135636-24b9-44e7-814a-af5872fd03a6" type="0">
            <message>  Copying file from "D:\home\site\repository\build\Compile\Owin.dll" to "D:\home\site\repository\build\Compile\_PublishedWebsites\Lucene\bin\Owin.dll".</message>
        </entry>
```

This makes the build process much slower than usual, in my case around 1h30, because I assume the file is flushed on the disk every seconds or so, and when the file reaches 1MB then most of the time is spent writing on disk.

I would suggest having an option to use a simple append only text file at least as the XML structure makes it very slow in this case.

@sebastienros I tried deploying https://github.com/OrchardCMS/Orchard branch 1.9.x, but that failed early (*error MSB4057: The target "PreComputeCompileTypeScript" does not exist in the project.*). What is the right way to test your scenario?

--- a/Kudu.Contracts/Deployment/IDetailedLogger.cs
+++ b/Kudu.Contracts/Deployment/IDetailedLogger.cs
@@ -1,0 +1,15 @@
+﻿using Kudu.Core.Deployment;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kudu.Core.Deployment
+{
+    public interface IDetailedLogger : ILogger
+    {
+        IEnumerable<LogEntry> GetLogEntries();
+        IEnumerable<LogEntry> GetLogEntryDetails(string entryId);
+    }
+}

--- a/Kudu.Contracts/Kudu.Contracts.csproj
+++ b/Kudu.Contracts/Kudu.Contracts.csproj
@@ -106,6 +106,7 @@
     <Compile Include="Deployment\IDeploymentManagerFactory.cs" />
     <Compile Include="Deployment\IDeploymentStatusManager.cs" />
     <Compile Include="Deployment\IDeploymentStatusFile.cs" />
+    <Compile Include="Deployment\IDetailedLogger.cs" />
     <Compile Include="Deployment\ILogger.cs" />
     <Compile Include="Deployment\LogEntry.cs" />
     <Compile Include="Deployment\LogEntryType.cs" />

--- a/Kudu.Core.Test/Kudu.Core.Test.csproj
+++ b/Kudu.Core.Test/Kudu.Core.Test.csproj
@@ -119,6 +119,8 @@
     <Compile Include="Settings\DeploymentSettingsPrioritiesTests.cs" />
     <Compile Include="Settings\SettingsProvidersTests.cs" />
     <Compile Include="SSHKeyManagerFacts.cs" />
+    <Compile Include="StructuredTextDocumentFacts.cs" />
+    <Compile Include="StructuredTextLoggerFacts.cs" />
     <Compile Include="Tracing\XmlTracerTests.cs" />
     <Compile Include="Tracing\SiteExtensionLogManagerTests.cs" />
     <Compile Include="TraceModuleFacts.cs" />

--- a/Kudu.Core.Test/MockFileSystem.cs
+++ b/Kudu.Core.Test/MockFileSystem.cs
@@ -49,6 +49,11 @@ namespace Kudu.Core.Test
                             return reader.ReadToEnd();
                         }
                     });
+            fileBase.Setup(f => f.ReadAllLines(It.IsAny<string>()))
+                    .Returns((string path) =>
+                    {
+                        return fileBase.Object.ReadAllText(path).Split(new[] { System.Environment.NewLine }, StringSplitOptions.None);
+                    });
 
             dirInfoFactory.Setup(d => d.FromDirectoryName(It.IsAny<string>()))
                           .Returns(dirInfoBase.Object);

--- a/Kudu.Core.Test/StructuredTextDocumentFacts.cs
+++ b/Kudu.Core.Test/StructuredTextDocumentFacts.cs
@@ -1,0 +1,107 @@
+﻿using Kudu.Core.Deployment;
+using Kudu.Core.Infrastructure;
+using Moq;
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using Xunit;
+
+namespace Kudu.Core.Test
+{
+    public class StructuredTextDocumentFacts
+    {
+        public T Id<T>(T obj)
+        {
+            return obj;
+        }
+
+        [Theory]
+        [InlineData("\tInvalid log entry that starts with \\t")]
+        [InlineData("Invalid log entry \n that contains a \\n")]
+        public void StructuredDocumentThrowsAFormatExceptionForInvalidLogEntries(string logEntry)
+        {
+            //Arrange
+            Exception exception = null;
+            var structuredDocument = new StructuredTextDocument<string>(@"x:\deployment\log.log", Id, Id);
+
+            //Act
+            try
+            {
+                structuredDocument.Write(logEntry, 0);
+            }
+            catch (Exception e)
+            {
+                exception = e;
+            }
+
+            //Assert
+            Assert.IsType(typeof(FormatException), exception);
+            Assert.Contains("Serialized log of type", exception.Message);
+        }
+
+        [Theory]
+        [InlineData("log entry 1", 0)]
+        [InlineData("log entry 2", 5)]
+        [InlineData("log entry 3", 11)]
+        public void StructuredDocumentLogsWithCorrectDepth(string logEntry, int depth)
+        {
+            //Arrange
+            var fileSystem = new Mock<IFileSystem>();
+            var fileBase = new Mock<FileBase>();
+            var stream = new TestMemoryStream();
+            var structuredDocument = new StructuredTextDocument<string>(@"x:\log.log", Id, Id);
+            fileSystem.Setup(f => f.File).Returns(fileBase.Object);
+
+            fileBase.Setup(f => f.Open(It.IsAny<string>(), It.IsAny<FileMode>(), It.IsAny<FileAccess>(), It.IsAny<FileShare>()))
+                    .Returns(stream);
+            FileSystemHelpers.Instance = fileSystem.Object;
+
+            //Act
+            structuredDocument.Write(logEntry, depth);
+
+            //Assert
+            Assert.True(stream.Position > 0, "must write data");
+            stream.Position = 0;
+            var result = new StreamReader(stream).ReadToEnd();
+            var writtenDepth = StructuredTextDocument.GetEntryDepth(result);
+            Assert.Equal(depth, writtenDepth);
+            Assert.Equal(string.Concat(logEntry, System.Environment.NewLine), result.Substring(writtenDepth));
+            stream.RealDispose();
+        }
+
+        [Fact]
+        public void StructuredDocumentParsingTest()
+        {
+            //Arrange
+            FileSystemHelpers.Instance = MockFileSystem.GetMockFileSystem(@"x:\log.log", () => "Message 1\r\nmessage 2\r\nmessage 3\r\n\tsub message 3.1\r\n\tsub message 3.2\r\n\tsub message 3.3\r\nmessage 4\r\n\tsub message 4.1");
+            var ExpectedContent = new int[] { 0, 0, 3 , 1 };
+            var structuredDocument = new StructuredTextDocument<string>(@"x:\log.log", Id, Id);
+
+            //Act
+            var parsed = structuredDocument.GetDocument().ToArray();
+
+            //Assert
+            Assert.Equal(4, parsed.Length);
+            for (var i = 0; i < parsed.Count(); i++)
+            {
+                Assert.Equal(ExpectedContent[i], parsed[i].Children.Count());
+            }
+        }
+
+        internal class TestMemoryStream : MemoryStream
+        {
+            public void RealDispose()
+            {
+                base.Dispose();
+            }
+
+            [SuppressMessage("Microsoft.Usage", "CA2215", Justification = "We need the stream to still be readable after using statement")] 
+            protected override void Dispose(bool disposing)
+            {
+                //no-op for testing
+            }
+        }
+    }
+}

--- a/Kudu.Core.Test/StructuredTextLoggerFacts.cs
+++ b/Kudu.Core.Test/StructuredTextLoggerFacts.cs
@@ -1,0 +1,80 @@
+﻿using Kudu.Core.Deployment;
+using Kudu.Core.Infrastructure;
+using Kudu.Core.Tracing;
+using Moq;
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using Xunit;
+
+namespace Kudu.Core.Test
+{
+    public class StructuredTextLoggerFacts
+    {
+        [Theory]
+        [InlineData("This message is clean from any illegal characters", LogEntryType.Message)]
+        [InlineData("This message has \"double quotes\" in it.", LogEntryType.Error)]
+        [InlineData("This message has \n new lines \r\n in it", LogEntryType.Warning)]
+        public void StructuredTextLoggerLogFunction(string message, LogEntryType type)
+        {
+            //Arrange
+            var fileSystem = new Mock<IFileSystem>();
+            var fileBase = new Mock<FileBase>();
+            var stream = new TestMemoryStream();
+            var analytics = new Mock<IAnalytics>();
+            var logger = new StructuredTextLogger(@"x:\log.log", analytics.Object);
+
+            fileSystem.Setup(f => f.File).Returns(fileBase.Object);
+
+            fileBase.Setup(f => f.Open(It.IsAny<string>(), It.IsAny<FileMode>(), It.IsAny<FileAccess>(), It.IsAny<FileShare>()))
+                    .Returns(stream);
+            fileBase.Setup(f => f.OpenRead(It.IsAny<string>()))
+                    .Returns((string path) => 
+                    {
+                        stream.Position = 0;
+                        return stream;
+                    });
+            fileBase.Setup(f => f.ReadAllText(It.IsAny<string>()))
+                    .Returns((string path) =>
+                    {
+                        using (var reader = new StreamReader(fileBase.Object.OpenRead(path)))
+                        {
+                            return reader.ReadToEnd();
+                        }
+                    });
+            fileBase.Setup(f => f.ReadAllLines(It.IsAny<string>()))
+                    .Returns((string path) =>
+                    {
+                        return fileBase.Object.ReadAllText(path).Split(new[] { System.Environment.NewLine }, StringSplitOptions.None);
+                    });
+
+            FileSystemHelpers.Instance = fileSystem.Object;
+
+            //Act
+            logger.Log(message, type);
+            var entries = logger.GetLogEntries();
+
+            //Assert
+            Assert.Equal(1, entries.Count());
+            Assert.Equal(message, entries.First().Message);
+            Assert.Equal(type, entries.First().Type);
+            stream.RealDispose();
+        }
+
+        internal class TestMemoryStream : MemoryStream
+        {
+            public void RealDispose()
+            {
+                base.Dispose();
+            }
+
+            [SuppressMessage("Microsoft.Usage", "CA2215", Justification = "We need the stream to still be readable after using statement")] 
+            protected override void Dispose(bool disposing)
+            {
+                //no-op for testing
+            }
+        }
+    }
+}

--- a/Kudu.Core/Deployment/DeploymentManager.cs
+++ b/Kudu.Core/Deployment/DeploymentManager.cs
@@ -33,7 +33,8 @@ namespace Kudu.Core.Deployment
         private readonly IDeploymentStatusManager _status;
         private readonly IWebHooksManager _hooksManager;
 
-        private const string LogFile = "log.xml";
+        private const string XmlLogFile = "log.xml";
+        private const string TextLogFile = "log.log";
         private const string TemporaryDeploymentIdPrefix = "temp-";
         public const int MaxSuccessDeploymentResults = 10;
 
@@ -94,7 +95,7 @@ namespace Kudu.Core.Deployment
 
                 VerifyDeployment(id, IsDeploying);
 
-                var logger = new XmlLogger(path, _analytics);
+                var logger = GetLoggerForFile(path);
                 List<LogEntry> entries = logger.GetLogEntries().ToList();
 
                 // Determine if there's details to show at all
@@ -121,7 +122,7 @@ namespace Kudu.Core.Deployment
 
                 VerifyDeployment(id, IsDeploying);
 
-                var logger = new XmlLogger(path, _analytics);
+                var logger = GetLoggerForFile(path);
 
                 return logger.GetLogEntryDetails(entryId).ToList();
             }
@@ -792,8 +793,8 @@ namespace Kudu.Core.Deployment
         public ILogger GetLogger(string id)
         {
             var path = GetLogPath(id);
-            var xmlLogger = new XmlLogger(path, _analytics);
-            return new ProgressLogger(id, _status, new CascadeLogger(xmlLogger, _globalLogger));
+            var logger = GetLoggerForFile(path);
+            return new ProgressLogger(id, _status, new CascadeLogger(logger, _globalLogger));
         }
 
         /// <summary>
@@ -834,9 +835,21 @@ namespace Kudu.Core.Deployment
             return Path.Combine(GetRoot(id), Constants.ManifestFileName);
         }
 
+        /// <summary>
+        /// This function handles getting the path for the log file.
+        /// If 'log.xml' exists then use that which will use XmlLogger, this is to support existing log files
+        /// else use 'log.log' which will use TextLogger. Moving forward deployment will always use text logger.
+        /// The logic to get the right logger is in GetLoggerForFile()
+        /// </summary>
+        /// <param name="id">deploymentId which is part of the path for the log file</param>
+        /// <param name="ensureDirectory">Create the directory if it doesn't exist</param>
+        /// <returns>log file path</returns>
         private string GetLogPath(string id, bool ensureDirectory = true)
         {
-            return Path.Combine(GetRoot(id, ensureDirectory), LogFile);
+            var logPath = Path.Combine(GetRoot(id, ensureDirectory), XmlLogFile);
+            return FileSystemHelpers.FileExists(logPath)
+                ? logPath
+                : Path.Combine(GetRoot(id, ensureDirectory), TextLogFile);
         }
 
         private string GetRoot(string id, bool ensureDirectory = true)
@@ -854,6 +867,24 @@ namespace Kudu.Core.Deployment
         private bool IsActive(string id)
         {
             return id.Equals(_status.ActiveDeploymentId, StringComparison.OrdinalIgnoreCase);
+        }
+
+        /// <summary>
+        /// use XmlLogger if the file ends with .xml
+        /// Otherwise use StructuredTextLogger
+        /// </summary>
+        /// <param name="logPath"></param>
+        /// <returns>XmlLogger or StructuredTextLogger</returns>
+        private IDetailedLogger GetLoggerForFile(string logPath)
+        {
+            if (logPath.EndsWith(".xml", StringComparison.OrdinalIgnoreCase))
+            {
+                return new XmlLogger(logPath, _analytics);
+            }
+            else
+            {
+                return new StructuredTextLogger(logPath, _analytics);
+            }
         }
 
         private class DeploymentAnalytics : IDisposable

--- a/Kudu.Core/Deployment/StructuredTextDocument.cs
+++ b/Kudu.Core/Deployment/StructuredTextDocument.cs
@@ -1,0 +1,129 @@
+﻿using Kudu.Core.Infrastructure;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Kudu.Core.Deployment
+{
+    public class StructuredTextDocument<T>
+    {
+        private const string LogTemplate = "{0}{1}";
+        private readonly string _path;
+        private readonly Func<T, string> _serializer;
+        private readonly Func<string, T> _deserializer;
+
+        public StructuredTextDocument(string path, Func<T, string> serializer, Func<string, T> deserializer)
+        {
+            _path = path;
+            _serializer = serializer;
+            _deserializer = deserializer;
+        }
+
+        public IEnumerable<StructuredTextDocumentEntry<T>> GetDocument()
+        {
+            return ParseLines(FileSystemHelpers.ReadAllLines(_path), startIndex: 0, depth: 0);
+        }
+
+        /// <summary>
+        /// Write uses the serializer that was specified in the constructor to convert T into a string
+        /// then logs it with the correct depth.
+        /// The depth determines how many \t are there in front of the log message. Which implies the
+        /// Parent/Child relationship between the log entries.
+        /// Here is an example of how the log looks like with 1 level of depth
+        /// 0000-00-00T00:00:00.0000000Z,Message 1,00000000-0000-0000-0000-000000000000,0
+        /// 0000-00-00T00:00:00.0000000Z,Message 2,00000000-0000-0000-0000-000000000000,0
+        /// \t0000-00-00T00:00:00.0000000Z,Message 2.1,,0
+        /// \t0000-00-00T00:00:00.0000000Z,Message 2.2,,0
+        /// 0000-00-00T00:00:00.0000000Z,Message 3,00000000-0000-0000-0000-000000000000,0
+        /// </summary>
+        /// <param name="value">Object to be serialized and logged in the StructuredTextDocument</param>
+        /// <param name="depth">Determines the Parent/Child relationship between log entries. depth = 0 is parent for depth = 1</param>
+        public void Write(T value, int depth)
+        {
+            var stringValue = _serializer(value);
+            ValidateStringMessage(stringValue);
+
+            var stringDepth = string.Concat(Enumerable.Range(0, depth).Select(_ => StructuredTextDocument.LeadingCharacter));
+            var logEntry = string.Format(LogTemplate, stringDepth, stringValue);
+            Log(logEntry);
+        }
+
+        private IEnumerable<StructuredTextDocumentEntry<T>> ParseLines(string[] lines, int startIndex, int depth)
+        {
+            List<StructuredTextDocumentEntry<T>> collection = new List<StructuredTextDocumentEntry<T>>();
+            StructuredTextDocumentEntry<T> logEntry = null;
+            for (var i = startIndex; i < lines.Length; i++)
+            {
+                //skip empty lines
+                if (string.IsNullOrWhiteSpace(lines[i]))
+                    continue;
+
+                var entryDepth = StructuredTextDocument.GetEntryDepth(lines[i]);
+                if (entryDepth == depth)
+                {
+                    logEntry = new StructuredTextDocumentEntry<T>
+                    {
+                        LogEntry = _deserializer(lines[i].Substring(depth)),
+                        Children = Enumerable.Empty<StructuredTextDocumentEntry<T>>()
+                    };
+                    collection.Add(logEntry);
+                }
+                else if (entryDepth > depth)
+                {
+                    if (logEntry == null)
+                    {
+                        throw new FormatException(string.Format("log file '{0}' is in an invalid format", _path));
+                    }
+
+                    logEntry.Children = ParseLines(lines, i, entryDepth);
+                    i += (logEntry.Children.Count() - 1);
+                }
+                else if (entryDepth < depth)
+                {
+                    return collection;
+                }
+            }
+            return collection;
+        }
+
+        private void Log(string logEntry)
+        {
+            FileSystemHelpers.AppendAllTextToFile(_path, string.Concat(logEntry, System.Environment.NewLine));
+        }
+
+        private static void ValidateStringMessage(string message)
+        {
+            if (message.StartsWith(StructuredTextDocument.LeadingCharacter, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new FormatException(
+                    string.Format("Serialized log of type '{0}' => '{1}' can't start with character '{2}'",
+                    typeof(T).Name,
+                    message,
+                    StructuredTextDocument.PrintableLeadingCharacter));
+            }
+
+            foreach (var pair in StructuredTextDocument.NotAllowedSequances)
+            {
+                if (message.IndexOf(pair.Key, StringComparison.OrdinalIgnoreCase) != -1)
+                {
+                    throw new FormatException(
+                        string.Format("Serialized log of type '{0}' => '{1}' contains an invalid sequance '{2}' for a logEntry in a StructuredTextDocument",
+                        typeof(T).Name,
+                        message,
+                        pair.Value));
+                }
+            }
+        }
+    }
+
+    public static class StructuredTextDocument
+    {
+        public const string LeadingCharacter = "\t";
+        public const string PrintableLeadingCharacter = "\\t";
+        public static readonly IDictionary<string, string> NotAllowedSequances = new Dictionary<string, string> {  { "\r\n", "\\r\\n" }, { "\r", "\\r" }, { "\n", "\\n" }};
+        public static int GetEntryDepth(string entry)
+        {
+            return entry.TakeWhile(c => LeadingCharacter.Equals(c.ToString())).Count();
+        }
+    }
+}

--- a/Kudu.Core/Deployment/StructuredTextDocumentEntry.cs
+++ b/Kudu.Core/Deployment/StructuredTextDocumentEntry.cs
@@ -1,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kudu.Core.Deployment
+{
+    public class StructuredTextDocumentEntry<T>
+    {
+        public T LogEntry { get; set; }
+        public IEnumerable<StructuredTextDocumentEntry<T>> Children { get; set; }
+    }
+}

--- a/Kudu.Core/Deployment/StructuredTextLogger.cs
+++ b/Kudu.Core/Deployment/StructuredTextLogger.cs
@@ -1,0 +1,127 @@
+﻿using Kudu.Core.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Kudu.Core.Deployment
+{
+    public class StructuredTextLogger : IDetailedLogger
+    {
+        private readonly static object DocumentLock = new object();
+        private readonly IAnalytics _analytics;
+        private int _depth;
+        private readonly string _path;
+        private const string LogEntrySeparator = ",";
+        private readonly StructuredTextDocument<LogEntry> _structuredTextDocument;
+
+        private static IEnumerable<KeyValuePair<string, string>> EscapeChars = new Dictionary<string, string>
+        {
+            { ",", "&comma;" }
+        }
+        .Union(StructuredTextDocument.NotAllowedSequances);
+
+        public StructuredTextLogger(string path, IAnalytics analytics)
+        {
+            _depth = 0;
+            _path = path;
+            _analytics = analytics;
+            _structuredTextDocument = new StructuredTextDocument<LogEntry>(path,
+                // DateTime.ToString("o") => "2015-08-04T00:08:38.5489308Z"
+                e => string.Join(LogEntrySeparator, e.LogTime.ToString("o"), e.Message, e.Id, (int)e.Type),
+                str =>
+                {
+                    var splitted = str.Split(new[] { LogEntrySeparator }, StringSplitOptions.None);
+                    if (splitted.Length == 4)
+                    {
+                        var time = DateTime.Parse(splitted[0]).ToUniversalTime();
+                        var message = UnsanitizeValue(splitted[1]);
+                        var id = splitted[2];
+                        var type = (LogEntryType)Int32.Parse(splitted[3]);
+                        return new LogEntry(time, id, message, type);
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format("the log line \"{0}\" is in an invalid format", str));
+                    }
+                });
+        }
+
+        public ILogger Log(string value, LogEntryType type)
+        {
+            try
+            {
+                value = SanitizeValue(value);
+                lock(DocumentLock)
+                {
+                    var logId = _depth == 0 ? Guid.NewGuid().ToString() : string.Empty;
+                    _structuredTextDocument.Write(new LogEntry(DateTime.UtcNow, logId, value, type), _depth);
+                }
+
+                return new StructuredTextLogger(_path, _analytics)
+                {
+                    _depth = _depth + 1
+                };
+            }
+            catch(Exception e)
+            {
+                _analytics.UnexpectedException(e);
+                throw;
+            }
+        }
+
+        public IEnumerable<LogEntry> GetLogEntries()
+        {
+            try
+            {
+                return _structuredTextDocument.GetDocument().Select(e => e.LogEntry);
+            }
+            catch(Exception e)
+            {
+                _analytics.UnexpectedException(e);
+                throw;
+            }
+        }
+
+        public IEnumerable<LogEntry> GetLogEntryDetails(string entryId)
+        {
+            try
+            {
+                var entry = _structuredTextDocument.GetDocument().FirstOrDefault(s => s.LogEntry.Id.Equals(entryId, StringComparison.OrdinalIgnoreCase));
+                return entry == null
+                    ? Enumerable.Empty<LogEntry>()
+                    : entry.Children.Select(e => e.LogEntry);
+            }
+            catch(Exception e)
+            {
+                _analytics.UnexpectedException(e);
+                throw;
+            }
+        }
+
+        private static string SanitizeValue(string value)
+        {
+            foreach (var pair in EscapeChars)
+            {
+                if (value.Contains(pair.Key))
+                {
+                    value = value.Replace(pair.Key, pair.Value);
+                }
+            }
+
+            return value;
+        }
+
+        private static string UnsanitizeValue(string value)
+        {
+            foreach (var pair in EscapeChars)
+            {
+                if (value.Contains(pair.Value))
+                {
+                    value = value.Replace(pair.Value, pair.Key);
+                }
+            }
+
+            return value;
+        }
+    }
+}

--- a/Kudu.Core/Deployment/XmlLogger.cs
+++ b/Kudu.Core/Deployment/XmlLogger.cs
@@ -9,7 +9,7 @@ using Kudu.Core.Tracing;
 
 namespace Kudu.Core.Deployment
 {
-    public class XmlLogger : ILogger
+    public class XmlLogger : IDetailedLogger
     {
         private readonly static object LogLock = new object();
         private readonly string _path;

--- a/Kudu.Core/Infrastructure/FileSystemHelpers.cs
+++ b/Kudu.Core/Infrastructure/FileSystemHelpers.cs
@@ -96,6 +96,11 @@ namespace Kudu.Core.Infrastructure
             return Instance.File.ReadAllText(path);
         }
 
+        public static string[] ReadAllLines(string path)
+        {
+            return Instance.File.ReadAllLines(path);
+        }
+
         /// <summary>
         /// Replaces File.ReadAllText,
         /// Will do the same thing only this can work on files that are already open (and share read/write).

--- a/Kudu.Core/Kudu.Core.csproj
+++ b/Kudu.Core/Kudu.Core.csproj
@@ -147,6 +147,9 @@
     <Compile Include="Deployment\Generator\NodeSiteBuilder.cs" />
     <Compile Include="Deployment\Generator\NodeSiteEnabler.cs" />
     <Compile Include="Deployment\Generator\SiteBuilderFactory.cs" />
+    <Compile Include="Deployment\StructuredTextDocument.cs" />
+    <Compile Include="Deployment\StructuredTextDocumentEntry.cs" />
+    <Compile Include="Deployment\StructuredTextLogger.cs" />
     <Compile Include="Hooks\WebHooksManager.cs" />
     <Compile Include="Infrastructure\AspNet5Helper.cs" />
     <Compile Include="Infrastructure\InstanceIdUtility.cs" />

